src/stats_generator.rs

```diff
@@ -0,0 +1,270 @@
+use std::cmp::min;
+use std::ops::Add;
+
+use regex::Regex;
+use serde::Serialize;
+
+use crate::history::History;
+
+#[derive(Debug)]
+pub struct StatsGenerator<'a> {
+    history: &'a History,
+}
+
+#[derive(Debug, Clone, Serialize)]
+struct StatItem {
+    cmd: String,
+    count: i64,
+}
+
+impl<'a> StatsGenerator<'a> {
+    pub fn generate_stats(&self, limit: i16, command_limit: Option<i16>) -> String {
+        let mut lines = "".to_owned();
+        let count_history = Self::count_commands_from_db_history(self);
+        if count_history == 0 {
+            return "No history, no stats!".to_string();
+        }
+        lines.push_str("📊 Quick stats:\n");
+        lines.push_str(format!("  - history has {:?} items ;\n", count_history).as_mut_str());
+        let most_used_commands = self.most_used_commands(&limit);
+        lines.push_str(&Self::generate_command_stats(
+            self,
+            limit,
+            command_limit,
+            most_used_commands,
+        ));
+        lines
+    }
+
+    fn generate_command_stats(
+        &self,
+        limit: i16,
+        command_limit: Option<i16>,
+        stats: Vec<StatItem>,
+    ) -> String {
+        let mut lines = "".to_owned();
+        if !stats.is_empty() {
+            lines.push_str(
+                format!(
+                    "  - {:?} first commands, sorted by occurrences:\n",
+                    min(limit, stats.len() as i16)
+                )
+                .as_mut_str(),
+            );
+            let re = Regex::new("^(.*)/(.*)$").unwrap();
+            for item in &stats[..min(limit as usize, stats.len())] {
+                let cmd = item.clone().cmd;
+                if !cmd.trim().is_empty() {
+                    if command_limit.is_some() && cmd.len() < command_limit.unwrap() as usize {
+                        continue;
+                    }
+                    let relative_cmd = re.captures(&cmd).map(|dir_and_cmd| {
+                        format!(
+                            "{0} ({1})",
+                            dir_and_cmd.get(2).unwrap().as_str(),
+                            dir_and_cmd.get(1).unwrap().as_str()
+                        )
+                    });
+                    lines = lines.add(&*format!(
+                        "    {:#} ({:?})\n",
+                        relative_cmd.unwrap_or(cmd),
+                        item.count
+                    ));
+                }
+            }
+        }
+        lines
+    }
+
+    fn most_used_commands(&self, limit: &i16) -> Vec<StatItem> {
+        self
+            .history
+            .run_query("select substr(cmd,1,instr(cmd,' ')-1), count(1) as n from commands group by 1 order by 2 desc limit :limit ", &[
```

Comment:
You can probably simultaneously get top directories and top commands in those directories. Something like:

```
let current_dir = "your_current_directory_here"; // replace with your current directory

let sql = format!(r#"
WITH top_dirs AS (
    SELECT dir
    FROM commands
    WHERE dir IS NOT NULL
    GROUP BY dir
    ORDER BY COUNT(*) DESC
    LIMIT 3
),
dirs AS (
    SELECT dir FROM top_dirs
    UNION ALL SELECT :current_dir
)
SELECT dir, cmd
FROM commands
WHERE dir IN dirs
GROUP BY dir, cmd
ORDER BY COUNT(*) DESC
LIMIT 10
"#);

let mut stmt = self.connection.prepare(&sql).unwrap();
let rows = stmt.query_map(named_params!{":current_dir": &current_dir}, |row| {
    Ok((row.get(0)?, row.get(1)?))
}).unwrap();

for row in rows {
    let (dir, cmd): (String, String) = row.unwrap();
    println!("dir: {}, cmd: {}", dir, cmd);
}
```

(I didn't test the above.)

Reply:
It seems interesting but I think it would make the "stats" command less general, because it would be related to the current directory. Maybe we could introduce this as an option, later (I mean, in an other pull request)?